### PR TITLE
Qt usb telnet

### PR DESCRIPTION
**Pull request by Eryk-Mozdzen**

### CODE REVIEW

File: software/goose-visual/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+
+## Packages for Ubuntu 22.04
+
+```bash
+sudo apt-get install -y build-essential
```
Comment (delipl) [inline software/goose-visual/README.md]:
```suggestion
sudo apt-get install -y build-essential qtbase5-dev qtdeclarative5-dev qtwayland5 qml-module-qtquick-controls qml-module-qtquick-controls2  libqt5serialport5-dev
```

File: software/goose-visual/src/communication/telnet.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+void Telnet::handleReadyRead() {
+    const QByteArray data = socket.readAll();
+
+	for(int i=0; i<data.size(); i++) {
+		transfer.consume(data[i]);
+
+		Transfer::FrameRX frame;
+
+		if(transfer.receive(frame))
```
Comment (delipl) [inline software/goose-visual/src/communication/telnet.cpp]: Missed brackets

File: software/goose-visual/src/communication/usb.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
+#include "usb.h"
+#include <QDebug>
+
+USB::USB(const QString port, QObject *parent) : QObject{parent} {
+
+	serial.setPortName(port);
+	serial.setBaudRate(115200);
+
+	connect(&serial, &QSerialPort::readyRead, this, &USB::handleReadyRead);
+    connect(&serial, &QSerialPort::errorOccurred, this, &USB::handleError);
```
Comment (delipl) [inline software/goose-visual/src/communication/usb.cpp]: missed tab

File: software/goose-visual/src/communication/usb.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	connect(&timer, &QTimer::timeout, this, &USB::handleTimeout);
+
+	if(!serial.open(QIODevice::OpenModeFlag::ReadWrite))
```
Comment (delipl) [inline software/goose-visual/src/communication/usb.cpp]: Missed brackets

File: software/goose-visual/src/communication/usb.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+void USB::handleReadyRead() {
+	const QByteArray data = serial.readAll();
+
+	for(int i=0; i<data.size(); i++) {
+		transfer.consume(data[i]);
+
+		Transfer::FrameRX frame;
+
+		if(transfer.receive(frame))
```
Comment (delipl) [inline software/goose-visual/src/communication/usb.cpp]: same brackets

File: software/goose-visual/src/communication/usb.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+			receive(frame);
+	}
+}
+
+void USB::handleError(QSerialPort::SerialPortError error) {
+	if(error==QSerialPort::SerialPortError::OpenError || error==QSerialPort::SerialPortError::ResourceError)
+		serial.close();
+
+	if(error!=QSerialPort::SerialPortError::NoError)
```
Comment (delipl) [inline software/goose-visual/src/communication/usb.cpp]: Missed brackets

File: software/goose-visual/src/interface/escape_codes.cpp [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#include "escape_codes.h"
+
+std::ostream & operator<<(std::ostream &stream, const EscapeCode &code) {
+	stream << static_cast<char>(27) << "[38;5;" << static_cast<int>(code) << "m";
```
Comment (delipl) [inline software/goose-visual/src/interface/escape_codes.cpp]: What is 27 magic number? Necessary to print colors?

**Review by delipl — APPROVED** [review]

File: software/goose-visual/src/interface/escape_codes.cpp (hunk 7 above)
Comment (Eryk-Mozdzen) [reply software/goose-visual/src/interface/escape_codes.cpp]: Escape ESC character, not printable

**Review by delipl — APPROVED** [review]

**Pull request merged**